refactor: Worker and ThreadSafeState

bartlomieju opened the pull request.

bartlomieju commented:
@ry are you gonna rewrite timers to use resources? That be very useful for this PR

ry commented:
Not yet - I will try tomorrow.

bartlomieju commented on cli/ops/workers.rs:
```diff
@@ -132,30 +140,31 @@ fn op_create_worker(
-
   if !has_source_code {
-    if let Some(module) = state.main_module() {
+    if let Some(module) = global_state.main_module() {
```
Workers are supposed to resolve entry script specifier relative to parent worker entry script (see `cli/tests/039_worker_deno_ns.ts`). That suggest that `main_module` should be stored per worker and not globally.

bartlomieju commented on cli/global_state.rs:
```diff
@@ -0,0 +1,309 @@
+pub struct ThreadSafeGlobalState(Arc<GlobalState>);
+
+#[cfg_attr(feature = "cargo-clippy", allow(stutter))]
+pub struct GlobalState {
```
s/GlobalState/ProgramState/ ?

bartlomieju commented:
Quick summary what happened in this PR:
1. Factor out fields from `ThreadSafeState` that are not "local" to isolate but are shared by whole program. Those fields were moved to new `ThreadSafeGlobalState` struct (the name needs to change)
2. Refactor `Worker` to hold reference to global state and keep its local state.
3. There is some churn here and there:
  - ops - added `global_state` argument, changed `state` to `global_state`
  - metrics - they are kept per isolate as well as "global" metrics - to be determined
4. This refactor opens a way to move resource table into `ThreadSafeState` and keep resources scoped per worker (so no sharing resources between workers, not sure if that's desirable but seems to be more secure, to be determined)
5. This refactor opens a way to implement changes described in #3116 - introducing distinct worker types depending on its role
6. This refactor opens a way to further refactor resources - worker resources are very tangled, and now can be refactored to `workerInternalChannels` and `workerExternalChannels` resources

ry commented on cli/global_state.rs:
```diff
@@ -0,0 +1,259 @@
+  }
+}
+
+impl Loader for ThreadSafeGlobalState {
```
I would have thought this is more of a per-worker thing...

bartlomieju replied:
That is a good question. I think that `Resolver` should be per-worker (because only "main worker" should have import maps applied (https://github.com/WICG/import-maps/issues/157) and `DENO_DIR`/`file_fetcher` are defined once.

bartlomieju replied:
I've changed my mind, it will be `Loader` trait that is implemented on the `ThreadSafeState` (so worker's state). And it will handle checking of permissions (so you can scope dynamic imports per worker :tada:) and eventually fetch module using file fetcher from global state :+1:

ry commented on cli/global_state.rs:
```diff
@@ -0,0 +1,259 @@
+use std::str;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
```
I would have expected a static global variable with the global state...

bartlomieju replied:
Yeah, makes sense

bartlomieju replied:
Might be hard to properly do this for tests with single global state but I'll give it a try anyway.

ry replied:
Yeah.. It's a good point. Even if you don't make it a static global, I don't think we need to modify the op parameters, rather just add something like `fn ThreadSafeState::global_state() -> ThreadSafeGlobalState`

bartlomieju replied:
I ended up with `global_state` field in `ThreadSafeState`.

ry commented on cli/ops/repl.rs:
```diff
@@ -11,15 +12,12 @@ use deno::*;
 use std::sync::Mutex;
 
-pub fn init(i: &mut Isolate, s: &ThreadSafeState) {
+pub fn init(i: &mut Isolate, gs: &ThreadSafeGlobalState, s: &ThreadSafeState) {
```
If the state is global, why does it need to be tediously passed around? A static global should be used to reference it.

bartlomieju replied:
Probably it doesn't - I'll rework to use static global

ry commented on core/modules.rs:
```diff
@@ -30,7 +30,7 @@ use std::sync::Mutex;
   dyn Future<Item = SourceCodeInfo, Error = ErrBox> + Send;
 
-pub trait Loader: Send + Sync {
+pub trait Resolver: Send + Sync {
```
Are these changes necessary? I'd rather have core changes done in a separate PR so they can be considered independently.

bartlomieju replied:
I believe so, please check [this comment](https://github.com/denoland/deno/pull/3242#discussion_r341152908)

bartlomieju commented:
> Quick summary what happened in this PR:
> 1. Factor out fields from `ThreadSafeState` that are not "local" to isolate but are shared by whole program. Those fields were moved to new `ThreadSafeGlobalState` struct (the name needs to change)
>2. Refactor `Worker` to hold reference to global state and keep its local state.
>3. There is some churn here and there:
>  - ops - added `global_state` argument, changed `state` to `global_state`
>  - metrics - they are kept per isolate as well as "global" metrics - to be determined
>4. This refactor opens a way to move resource table into `ThreadSafeState` and keep resources scoped per worker (so no sharing resources between workers, not sure if that's desirable but seems to be more secure, to be determined)
>5. This refactor opens a way to implement changes described in #3116 - introducing distinct worker types depending on its role
>6. This refactor opens a way to further refactor resources - worker resources are very tangled, and now can be refactored to `workerInternalChannels` and `workerExternalChannels` resources

Update after recent changes:
2. Worker now holds its local state and that state holds a reference to global state.
3. Removed this churn due to refactor from point 2.
6. I went all in and refactored "worker" resource. It is now defined in `cli/worker.rs` and is a struct holding references to internal and external channels of the worker.

bartlomieju commented:
@ry @piscisaureus please review

bartlomieju commented on cli/global_state.rs:
```diff
@@ -0,0 +1,259 @@
+    .unwrap()
+  }
+
+  pub fn stateful_op<D>(
```
Remove

bartlomieju commented:
@ry re offline discussion: I had to change `Worker` resource a bit - only "external" channels are held in resource table and internal channels are still in `State.worker_channels`. Otherwise internal channels are never dropped and worker can't exit.

I think this can be cleaned up more but I don't want to tackle that in this PR, there's already a lot going on.

ry commented on cli/resources.rs:
```diff
@@ -372,78 +364,6 @@ pub fn add_reqwest_body(body: ReqwestDecoder) -> Resource {
-pub fn get_message_stream_from_worker(rid: ResourceId) -> WorkerReceiverStream {
-  WorkerReceiverStream { rid }
-}
-
```
👍

ry commented on cli/state.rs:
```diff
@@ -195,158 +163,74 @@ impl Loader for ThreadSafeState {
+    let import_map: Option<ImportMap> =
+      match global_state.flags.import_map_path.as_ref() {
+        None => None,
+        Some(file_path) => Some(ImportMap::load(file_path)?),
```
Does this mean the import map is loaded every time a worker is created? (I note this is the behavior beforehand too - just curious)

bartlomieju replied:
Yes, this bit is to be removed for "web workers". Only main worker should have import map applied. Will open an issue for that

ry commented on cli/worker.rs:
```diff
@@ -115,6 +143,37 @@ impl Worker {
     })
   }
+
+  /// Post message to worker as a host or privileged overlord
```
: )

ry commented on cli/worker.rs:
```diff
@@ -127,16 +186,36 @@ impl Future for Worker {
   }
 }
 
+pub struct WorkerReceiver {
```
This could use a /// comment

bartlomieju commented:
@ry

```
Refactor Worker and ThreadSafeState
* Split ThreadSafeState into State and GlobalState. State is a "local" state belonging to "Worker" while "GlobalState" is state shared by whole program.
* Update "Worker" and ops to use "GlobalState" where applicable
* Move and refactor "WorkerChannels" resource
```
